### PR TITLE
BackupDb streams data from database

### PR DESCRIPTION
**Pull request by Quetzacoalt91**
| Questions         | Answers
| ------------------| -------------------------------------------------------
| Description?      | Following an ugprade attempt made on a production like shop, we discovered the database backup takes too much time. It did not complete after a whole day of work while a save of the same database via PhpMyAdmin takes a couple of minutes. This PRs is a second attempt to see an improvement on the time needed to backup de database.
| Type?             | improvement
| BC breaks?        | Nope
| Deprecations?     | Nope
| Fixed ticket?     | /
| Sponsor company   | @PrestaShopCorp
| How to test?      | Backups are faster. You may compare the results with https://github.com/PrestaShop/SeamlessUpgradeToolbox

### CODE REVIEW

**Comment by Quetzacoalt91** [comment]
Restoring the files generated by this PR crashes the database with `ERROR 2006 (HY000) at line 27: Server has gone away`. I need to split the values in several `INSERT INTO` statements.

File: classes/Task/Upgrade/BackupDb.php [hunk 1]
```diff
@@ -39,6 +41,8 @@ class BackupDb extends AbstractTask
 {
     const TASK_TYPE = 'upgrade';
 
+    const MAX_SIZE_PER_INSERT_STMT = 950000;
+
     /**
      * @throws Exception
      */
```
Comment (M0rgan01) [inline classes/Task/Upgrade/BackupDb.php:50]: The run function still remains very hard to read... 219 lines long for a function is too much, even after extracting several private functions. 

It's a good start :)

File: classes/Task/Upgrade/BackupDb.php [hunk 2]
```diff
@@ -215,49 +162,71 @@ public function run(): int
...
-                    $this->container->getState()->setBackupLoopLimit($this->container->getState()->getBackupLoopLimit() + 200);
-                    $sizeof = $this->container->getDb()->numRows();
-                    if ($data && ($sizeof > 0)) {
-                        // Export the table data
+            if ($this->container->getState()->getBackupTable()) {
+                $backup_loop_limit = $this->container->getState()->getBackupLoopLimit();
+
+                $dbLink->setAttribute(PDO::MYSQL_ATTR_USE_BUFFERED_QUERY, false);
+                /** @see https://dev.mysql.com/doc/refman/8.4/en/select.html specifies a large LIMIT value to get the whole table */
+                $data = $dbLink->prepare('SELECT * FROM `' . $table . '` LIMIT ' . (int) $backup_loop_limit . ',18446744073709551615');
```
Comment (M0rgan01) [inline classes/Task/Upgrade/BackupDb.php:175]: it might be worth outputting 18446744073709551615 as a constant, with a variable name that could help us understand the number

File: classes/Task/Upgrade/BackupDb.php [hunk 3]
```diff
@@ -271,39 +240,134 @@ public function run(): int
...
+            $this->logger->info($this->translator->trans('%s tables have been saved.', [$numberOfSyncedTables]));
+        }
+
         if ($tablesToBackup->getRemainingTotal()) {
-            $this->logger->debug($this->translator->trans('%s tables have been saved.', [$found]));
             $this->next = 'backupDb';
             $this->stepDone = false;
-            $this->logger->info($this->translator->trans('Database backup: %s table(s) left...', [$tablesToBackup->getRemainingTotal()]));
+            if ($numberOfSyncedTables) {
+                $this->logger->info($this->translator->trans('Database backup: %s table(s) left...', [$tablesToBackup->getRemainingTotal() + (int) !empty($row)]));
```
Comment (M0rgan01) [inline classes/Task/Upgrade/BackupDb.php]: why did you add (int) !empty($row)? Was the value incorrect before?

File: classes/Task/Upgrade/BackupDb.php [hunk 4]
```diff
@@ -271,39 +240,134 @@ public function run(): int
...
+
+        if (empty($tablesToBackup)) {
+            throw (new UpgradeException($this->translator->trans('No valid tables were found to back up. Backup of database canceled.')))->setSeverity(UpgradeException::SEVERITY_ERROR);
         }
+
+        $tablesToBackup = new Backlog($listOfTables, count($listOfTables));
+
+        $this->container->getFileConfigurationStorage()->save($tablesToBackup->dump(), UpgradeFileNames::DB_TABLES_TO_BACKUP_LIST);
+
+        return ExitCode::SUCCESS;
```
Comment (M0rgan01) [inline classes/Task/Upgrade/BackupDb.php]: By mocking the database it should be possible to add unit tests on this function from hell... what do you think?

File: classes/Task/Upgrade/BackupDb.php (hunk 1 above)
Comment (Quetzacoalt91) [reply classes/Task/Upgrade/BackupDb.php:50]: Indeed. Unfortunately because of the local variable that can be modified inside the loops (such as the time spent, the quantity written in file...), I wasn't able to split everything without redoing the whole class.

File: classes/Task/Upgrade/BackupDb.php (hunk 2 above)
Comment (Quetzacoalt91) [reply classes/Task/Upgrade/BackupDb.php:175]: I though about it and preferred to put it here as 
* this is a value we won't change in the future, and
* it is the closest possible to its comment and the context in which it is used (the SQL request).

File: classes/Task/Upgrade/BackupDb.php (hunk 3 above)
Comment (Quetzacoalt91) [reply classes/Task/Upgrade/BackupDb.php]: I got errors from PHP complaining about adding boolean to an integer, but I can't reproduce it in a sandboxed env. I'll have another look at it and will remove it if possible.

File: classes/Task/Upgrade/BackupDb.php [hunk 5]
```diff
@@ -271,39 +240,134 @@ public function run(): int
...
 
-            return ExitCode::SUCCESS;
+            return true;
+        });
+
+        if (empty($tablesToBackup)) {
+            throw (new UpgradeException($this->translator->trans('No valid tables were found to back up. Backup of database canceled.')))->setSeverity(UpgradeException::SEVERITY_ERROR);
         }
+
+        $tablesToBackup = new Backlog($listOfTables, count($listOfTables));
```
Comment (Quetzacoalt91) [inline classes/Task/Upgrade/BackupDb.php]: There's an issue here, We're not taking the filtered list of tables.

### Dump with the PR
[auto-backupdb_000001_V8.0.5_20240925-180012-53fc0a60.txt](https://github.com/user-attachments/files/17135329/auto-backupdb_000001_V8.0.5_20240925-180012-53fc0a60.txt)

### Dump before the PR
[auto-backupdb_000001_V8.0.5_20240925-181548-6807d5a3.txt](https://github.com/user-attachments/files/17135335/auto-backupdb_000001_V8.0.5_20240925-181548-6807d5a3.txt)

File: classes/Task/Upgrade/BackupDb.php [hunk 6]
```diff
@@ -271,39 +240,146 @@ public function run(): int
...
+
+    private function isRemainingTimeEnough(int $elapsedTime): bool
+    {
+        $timeAllowed = (int) @ini_get('max_execution_time');
+
+        if ($timeAllowed <= 0) {
+            return true;
         }
+
+        return $elapsedTime < $timeAllowed - 5;
```
Comment (ga-devfront) [inline classes/Task/Upgrade/BackupDb.php:388]: I think it would me great to add a comment to explain the -5

**Review by M0rgan01 — DISMISSED** [review]
Great improvement

**Review by ga-devfront — DISMISSED** [review]
image

File: classes/Task/Upgrade/BackupDb.php (hunk 6 above)
Comment (Quetzacoalt91) [reply classes/Task/Upgrade/BackupDb.php:388]:
```suggestion
        // Remove 5 seconds on the allowed time to make sure we have time to save data and close files .
        return $elapsedTime < $timeAllowed - 5;
```

**Review by ga-devfront — DISMISSED** [review]

**Review by ga-devfront — DISMISSED** [review]

**Review by M0rgan01 — APPROVED** [review]

**Review by ga-devfront — APPROVED** [review]

**Review by AureRita — APPROVED** [review]
Hi @Quetzacoalt91 

Thank you for your PR, I tested it and it seems to works 

Tested from : 
1.7.8.8 to 8.2 
8.0.4 to 8.2  
8.2 to 9.0.0
8.0.4 to 9.0.0

Because the PR seems to works as expected, It's QA :heavy_check_mark: 

Thank you
